server/src/main/java/de/uftos/repositories/ucdl/parser/DefinitionParser.java

```diff
@@ -0,0 +1,691 @@
+          attributes.add(new ValueDto<>(UcdlToken.ATTRIBUTE, attribute));
+          attributeList = attributeList.jjtGetChild(0);
+        }
+        return new ElementDto(UcdlToken.ELEMENT, elementName, attributes, elementType);
+      }
+      case "ELEMENTNAME" -> {
+        String image = ((SimpleNode) root).jjtGetFirstToken().image;
+        if (!parameters.containsKey(image)) {
+          throw new ParseException("Parameter/Variable \"" + image + "\" does not exist!");
+        }
+        return new ValueDto<>(UcdlToken.VALUE_REFERENCE, image);
+      }
+      case "NUMBERELEMENT" -> {
+        if (((SimpleNode) root).jjtGetFirstToken().image.equals("size")) {
+          List<AbstractSyntaxTreeDto> set = new ArrayList<>();
+          set.add(buildAst(root.jjtGetChild(0), parameters));
+          return new OperatorDto(UcdlToken.SIZE, set);
+        } else {
+          return new ValueDto<>(UcdlToken.NUMBER, Integer.parseInt(
+              ((SimpleNode) root).jjtGetFirstToken().image));
+        }
+      }
+      case "SET" -> {
+        AbstractSyntaxTreeDto setName = buildAst(root.jjtGetChild(0).jjtGetChild(0), parameters);
+
+        ResourceType setType;
+
+        if (setName.getToken() == UcdlToken.NUMBER_SET) {
+          setType = ResourceType.NUMBER;
+        } else {
+          setType = ((ElementDto) setName).type();
+        }
+
+        List<AbstractSyntaxTreeDto> modifiers = new ArrayList<>();
+
+        Node modifierList = root.jjtGetChild(1);
+        while (modifierList.jjtGetNumChildren() > 0) {
+          if (modifierList.jjtGetNumChildren() == 2) {
+            SimpleNode flatmap = (SimpleNode) modifierList.jjtGetChild(0);
+            String attribute = flatmap.jjtGetFirstToken().image;
+            setType = getNextResourceType(setType, attribute);
+            modifiers.add(new ValueDto<>(UcdlToken.ATTRIBUTE, attribute));
+            modifierList = modifierList.jjtGetChild(1);
+
+          } else if (modifierList.jjtGetNumChildren() == 3) {
+            ResourceType thisType = parameters.get("this");
+            parameters.replace("this", setType);
+
+            List<AbstractSyntaxTreeDto> filters = new ArrayList<>();
+            filters.add(buildAst(modifierList.jjtGetChild(0), parameters));
+
+            Node filterList = modifierList.jjtGetChild(1);
+            while (filterList.jjtGetNumChildren() > 0) {
+              filters.add(buildAst(filterList.jjtGetChild(0), parameters));
+              filterList = filterList.jjtGetChild(1);
+            }
+
+            for (AbstractSyntaxTreeDto filter : filters) {
+              if ((filter.getToken() == UcdlToken.NUMBER_SET && setType != ResourceType.NUMBER)
+                  || (filter.getToken() == UcdlToken.RESOURCE_SET
+                  && ((SetDto) filter).type() != setType)) {
+                throw new ParseException("Sets can only be filtered using sets of the same type!");
+              }
```

Comment:
Use early `continue`s to make it readable!
Let's apply some logic...
Let
```
A := filter.getToken() == UcdlToken.NUMBER_SET
B := setType != ResourceType.NUMBER
C := filter.getToken() == UcdlToken.RESOURCE_SET
D := ((SetDto) filter).type() != setType
```
Then the expression becomes `(A && B) || (C && D)`. Negating this leaves us with `(NOT A || NOT B) && (NOT C || NOT D) = (NOT A && NOT C) || (NOT B && NOT C) || (NOT A && NOT D) || (NOT B && NOT D)`
Thus, if either of those terms is fulfilled we want to continue the loop and not throw an error.
This way, we can have four simple, easy to understand ifs without nesting of logic or ifs.